### PR TITLE
Independent disk new properties support added which are supported from API version 36.0

### PR DESCRIPTION
**Pull request by vbauzysvmware**
For: https://github.com/vmware/terraform-provider-vcd/pull/789
Dep: #383

Disk methods have now the ability to access new properties from API version 36.0. 
* DiskRecordType.SharingType
* DiskRecordType.UUID
* DiskRecordType.Encrypted
* Disk.SharingType
* Disk.UUID
* Disk.Encrypted

New method added
* Disk.GetAttachedVmsHrefs

`Vms.VmReference` changed to array to fix not correct deserialization

Tested with 10.2.2 and 10.3.1

### CODE REVIEW

File: .changes/v2.15.0/436-improvements.md [hunk 1]
```diff
@@ -0,0 +1 @@
+* Disk methods have now the ability to access new properties from API version 36.0. They are: `DiskRecordType.SharingType`, `DiskRecordType.UUID`, `DiskRecordType.Encrypted`, `Disk.SharingType`, `Disk.UUID` and `Disk.Encrypted`. Also, new method added `Disk.GetAttachedVmsHrefs`. `Vms.VmReference` changed to array to fix not correct deserialization [GH-436]
```
Comment (Didainius) [inline .changes/v2.15.0/436-improvements.md:1]:
```suggestion
* Disk methods have now the ability to access new properties from API version 36.0. They are: `DiskRecordType.SharingType`, `DiskRecordType.UUID`, `DiskRecordType.Encrypted`, `Disk.SharingType`, `Disk.UUID` and `Disk.Encrypted`. Also, new method added `Disk.GetAttachedVmsHrefs`. `Vms.VmReference` changed to array to fix incorrect deserialization [GH-436]
```

Comment (Didainius) [inline .changes/v2.15.0/436-improvements.md:1]: IMO the `Vms.VmReference changed to array to fix not correct deserialization` is also worth a `bugfix` entry for `VmReference`field. That automatically explains why a breaking change is done in structure + is easier to spot for someone looking bugfixes.

File: types/v56/types.go [hunk 2]
```diff
@@ -2661,8 +2661,8 @@ type Disk struct {
 	Description     string           `xml:"Description,omitempty"`
 	Files           *FilesList       `xml:"Files,omitempty"`
 	Link            []*Link          `xml:"Link,omitempty"`
-	Owner           *Owner           `xml:"Owner,omitempty"`
 	StorageProfile  *Reference       `xml:"StorageProfile,omitempty"`
+	Owner           *Owner           `xml:"Owner,omitempty"`
```
Comment (Didainius) [inline types/v56/types.go:2665]: Any reason for this field placement change?

File: govcd/disk.go [hunk 3]
```diff
@@ -369,8 +376,9 @@ func (vdc *Vdc) GetDiskByHref(diskHref string) (*Disk, error) {
 	util.Logger.Printf("[TRACE] Get Disk By Href: %s\n", diskHref)
 	Disk := NewDisk(vdc.client)
 
-	_, err := vdc.client.ExecuteRequest(diskHref, http.MethodGet,
-		"", "error retrieving Disk: %#v", nil, Disk.Disk)
+	_, err := vdc.client.ExecuteRequestWithApiVersion(diskHref, http.MethodGet,
+		"", "error retrieving Disk: %#v", nil, Disk.Disk,
```
Comment (Didainius) [inline govcd/disk.go:380]: Opportunity for fix
```suggestion
		"", "error retrieving Disk: %s", nil, Disk.Disk,
```

File: govcd/disk.go [hunk 4]
```diff
@@ -429,3 +437,50 @@ func (vdc *Vdc) GetDiskById(diskId string, refresh bool) (*Disk, error) {
...
+
+	if attachedVMsLink == nil {
+		return nil, fmt.Errorf("[ERROR] GetAttachedVmsHrefs - could not find request URL for attached vm in disk Link")
+	}
+
+	// Decode request
+	var vms = new(types.Vms)
+
+	_, err := disk.client.ExecuteRequest(attachedVMsLink.HREF, http.MethodGet,
+		attachedVMsLink.Type, "[ERROR] GetAttachedVmsHrefs - error getting attached vms: %s", nil, vms)
```
Comment (Didainius) [inline govcd/disk.go:471]:
```suggestion
		attachedVMsLink.Type, "[ERROR] GetAttachedVmsHrefs - error getting attached VMs: %s", nil, vms)
```

File: govcd/disk.go [hunk 5]
```diff
@@ -429,3 +437,50 @@ func (vdc *Vdc) GetDiskById(diskId string, refresh bool) (*Disk, error) {
...
+			util.Logger.Printf("[TRACE] GetAttachedVmsHrefs - found the proper link for request, HREF: %s, name: %s, type: %s,id: %s, rel: %s \n",
+				diskLink.HREF, diskLink.Name, diskLink.Type, diskLink.ID, diskLink.Rel)
+
+			attachedVMsLink = diskLink
+			break
+		}
+	}
+
+	if attachedVMsLink == nil {
+		return nil, fmt.Errorf("[ERROR] GetAttachedVmsHrefs - could not find request URL for attached vm in disk Link")
```
Comment (Didainius) [inline govcd/disk.go:464]: We usually used word `error` for errors. Not sure if this is just because it got confused with logging message or a specific reason.

**Review by Didainius — COMMENTED** [review]
A few comments and one more concern - could you actually test if these work with API V35.0 as default as we're soon going to merge it.

File: .changes/v2.15.0/436-improvements.md (hunk 1 above)
Comment (vbauzysvmware) [reply .changes/v2.15.0/436-improvements.md:1]: updated

File: types/v56/types.go (hunk 2 above)
Comment (vbauzysvmware) [reply types/v56/types.go:2665]: XML has to be in precise order otherwise serialization fails. Placement change - is the fix

File: .changes/v2.15.0/436-improvements.md (hunk 1 above)
Comment (vbauzysvmware) [reply .changes/v2.15.0/436-improvements.md:1]: moved to separate changelog bug fixes file

File: govcd/disk.go (hunk 3 above)
Comment (vbauzysvmware) [reply govcd/disk.go:380]: updated

File: govcd/disk.go (hunk 4 above)
Comment (vbauzysvmware) [reply govcd/disk.go:471]: updated

File: govcd/disk.go (hunk 5 above)
Comment (vbauzysvmware) [reply govcd/disk.go:464]: updated

**Comment by vbauzysvmware** [comment]
> 

Tested with 35.0 version and small fix added as error changed

File: types/v56/types.go [hunk 6]
```diff
@@ -2768,6 +2768,7 @@ type DiskRecordType struct {
 	BusSubType         string    `xml:"busSubType,attr,omitempty"`
 	BusTypeDesc        string    `xml:"busTypeDesc,attr,omitempty"`
 	AttachedVmCount    int32     `xml:"attachedVmCount,attr,omitempty"`
+	SharingType        string    `xml:"sharingType,attr,omitempty"`
```
Comment (lvirbalas) [inline types/v56/types.go:2771]: Just to confirm: is this field ignored in pre-v36.0 API versions?

Comment (lvirbalas) [reply types/v56/types.go:2771]: BTW, let's use the chance and remove the obsolete comments few lines above too (esp. now that we're moving away from V33.0)?

```
  // SizeB is not available in API V33.0. It is replaced by SizeMb
	//SizeB              int64   `xml:"sizeB,attr,omitempty"`
```

Comment (vbauzysvmware) [reply types/v56/types.go:2771]: not ignored:
cvc-complex-type.3.2.2: Attribute 'sharingType' is not allowed to appear in element 'Disk'.

XML is validate by XSD

Comment (vbauzysvmware) [reply types/v56/types.go:2769]: comment removed

Comment (lvirbalas) [reply types/v56/types.go:2769]: > not ignored:
> cvc-complex-type.3.2.2: Attribute 'sharingType' is not allowed to appear in element 'Disk'.
> 
> XML is validate by XSD

Interesting. But it is ignored by pre-V36.0 VCD versions?

Comment (vbauzysvmware) [reply types/v56/types.go:2769]: it's ignored if you fetch information, nothing will return for sharingType, but if you send on update or create you will get error which I showed.

Comment (lvirbalas) [reply types/v56/types.go:2769]: > but if you send on update or create you will get error which I showed.

Where did you show it?

After reading your reply, I don't understand: then how does independent disk update/create still work with previous VCD versions?

Comment (vbauzysvmware) [reply types/v56/types.go:2769]: error: cvc-complex-type.3.2.2: Attribute 'sharingType' is not allowed to appear in element 'Disk'.

If works because field sharingType is optional, old version API doesn't care about this field, new API cares if you send it

Comment (lvirbalas) [reply types/v56/types.go:2769]: > If works because field sharingType is optional, old version API doesn't care about this field, new API cares if you send it

And this is exactly what I wanted to know with my original question in this thread. Thanks!

File: .changes/v2.15.0/436-improvements.md [hunk 7]
```diff
@@ -0,0 +1 @@
+* Disk methods have now the ability to access new properties from API version 36.0. They are: `DiskRecordType.SharingType`, `DiskRecordType.UUID`, `DiskRecordType.Encrypted`, `Disk.SharingType`, `Disk.UUID` and `Disk.Encrypted`. Also, new method added `Disk.GetAttachedVmsHrefs [GH-436]
```
Comment (lvirbalas) [inline .changes/v2.15.0/436-improvements.md:1]: > Also, new method added `Disk.GetAttachedVmsHrefs [GH-436]

This new feature is unrelated to the new fields? If yes, it deserves a separate bullet point in the changelog.

Comment (vbauzysvmware) [reply .changes/v2.15.0/436-improvements.md]: changed

**Review by lvirbalas — APPROVED** [review]

**Review by adambarreiro — APPROVED** [review]

**Review by Didainius — APPROVED** [review]

**Review by mikeletux — APPROVED** [review]
LGTM!
